// @material/web/fab/lib/shared.js
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a;
import { __decorate } from "tslib";
import '../../elevation/elevation.js';
import '../../focus/focus-ring.js';
import '../../ripple/ripple.js';
import { html, LitElement, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { requestUpdateOnAriaChange } from '../../internal/aria/delegate.js';
// tslint:disable-next-line:enforce-comments-on-exported-symbols
export class SharedFab extends LitElement {
    constructor() {
        super(...arguments);
        /**
         * The size of the FAB.
         *
         * NOTE: Branded FABs cannot be sized to `small`, and Extended FABs do not
         * have different sizes.
         */
        this.size = 'medium';
        /**
         * The text to display on the FAB.
         */
        this.label = '';
        /**
         * Lowers the FAB's elevation.
         */
        this.lowered = false;
        /**
         * NOTE: For SSR use only as it will be overriden by icon slotchange event.
         *
         * Whether to display the icon or not in extended FAB. Does nothing on branded
         * and non-extended FABs.
         */
        this.hasIcon = false;
        /**
         * Lowers the FAB's elevation and places it into the `lowered` state.
         */
        this.reducedTouchTarget = false;
    }
    render() {
        // Needed for closure conformance
        const { ariaLabel } = this;
        return html `
      <button
        class="fab ${classMap(this.getRenderClasses())}"
        aria-label=${ariaLabel || nothing}
      >
        <md-elevation></md-elevation>
        <md-focus-ring></md-focus-ring>
        <md-ripple class="ripple"></md-ripple>
        ${this.renderTouchTarget()}
        ${this.renderIcon()}
        ${this.renderLabel()}
      </button>
    `;
    }
    getRenderClasses() {
        const isExtended = !!this.label;
        return {
            'lowered': this.lowered,
            'small': this.size === 'small' && !isExtended,
            'large': this.size === 'large' && !isExtended,
            'extended': isExtended,
            'hasIcon': !isExtended || this.hasIcon,
        };
    }
    renderTouchTarget() {
        return this.reducedTouchTarget ? html `` :
            html `<div class="touch-target"></div>`;
    }
    renderLabel() {
        return this.label ? html `<span class="label">${this.label}</span>` : '';
    }
    renderIcon() {
        const { ariaLabel } = this;
        return html `<span class="icon">
        <slot
            name="icon"
            aria-hidden=${ariaLabel || this.label ? 'true' : nothing}
            @slotchange=${this.onSlotchange}>
        </slot>
      </span>`;
    }
    onSlotchange(e) {
        const slotEl = e.target;
        const slottedEls = slotEl.assignedElements({ flatten: true });
        this.hasIcon = slottedEls.length !== 0;
    }
}
_a = SharedFab;
(() => {
    requestUpdateOnAriaChange(_a);
})();
/** @nocollapse */
SharedFab.shadowRootOptions = {
    mode: 'open',
    delegatesFocus: true,
};
__decorate([
    property()
], SharedFab.prototype, "size", void 0);
__decorate([
    property()
], SharedFab.prototype, "label", void 0);
__decorate([
    property({ type: Boolean })
], SharedFab.prototype, "lowered", void 0);
__decorate([
    property({ type: Boolean, attribute: 'has-icon' })
], SharedFab.prototype, "hasIcon", void 0);
__decorate([
    property({ type: Boolean, attribute: 'reduced-touch-target' })
], SharedFab.prototype, "reducedTouchTarget", void 0);
