// @material/web/chips/lib/chip.js
/**
 * @license
 * Copyright 2023 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */
var _a;
import { __decorate } from "tslib";
import '../../focus/focus-ring.js';
import '../../ripple/ripple.js';
import { html, LitElement, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { classMap } from 'lit/directives/class-map.js';
import { requestUpdateOnAriaChange } from '../../internal/aria/delegate.js';
/**
 * A chip component.
 */
export class Chip extends LitElement {
    constructor() {
        super(...arguments);
        this.disabled = false;
        this.label = '';
    }
    /**
     * Whether or not the primary ripple is disabled (defaults to `disabled`).
     * Some chip actions such as links cannot be disabled.
     */
    get rippleDisabled() {
        return this.disabled;
    }
    render() {
        return html `
      <div class="container ${classMap(this.getContainerClasses())}">
        ${this.renderOutline()}
        <md-focus-ring for=${this.primaryId}></md-focus-ring>
        <md-ripple for=${this.primaryId}
          ?disabled=${this.rippleDisabled}></md-ripple>
        ${this.renderPrimaryAction()}
        ${this.renderTrailingAction?.() || nothing}
      </div>
    `;
    }
    getContainerClasses() {
        return {
            disabled: this.disabled,
        };
    }
    renderContent() {
        return html `
      <span class="leading icon">
        ${this.renderLeadingIcon()}
      </span>
      <span class="label">${this.label}</span>
      <span class="touch"></span>
    `;
    }
    renderOutline() {
        return html `<span class="outline"></span>`;
    }
    renderLeadingIcon() {
        return html `<slot name="icon"></slot>`;
    }
}
_a = Chip;
(() => {
    requestUpdateOnAriaChange(_a);
})();
__decorate([
    property({ type: Boolean })
], Chip.prototype, "disabled", void 0);
__decorate([
    property()
], Chip.prototype, "label", void 0);
